api/v1beta1/machinedrainrules_types.go

```diff
@@ -0,0 +1,213 @@
+	//      - key: cluster.x-k8s.io/control-plane
+	//        operator: Exists
+	//  - selector:
+	//      matchLabels:
+	//        os: linux
+	//    clusterSelector:
+	//      matchExpressions:
+	//      - key: stage
+	//        operator: In
+	//        values:
+	//        - production
+	//
+	// +required
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:MaxItems=32
+	Machines []MachineDrainRuleMachineSelector `json:"machines"`
+
+	// pods defines to which Pods this MachineDrainRule should be applied.
+	//
+	// If pods is empty, the MachineDrainRule applies to all Pods in all Namespaces.
+	// If pods contains multiple selectors, the results are ORed.
+	// Within a single Pod selector the results of selector and namespaceSelector are ANDed.
+	// Pods will be selected from all Namespaces unless otherwise
+	// restricted with the namespaceSelector.
+	// An empty Pod selector matches all Pods in all Namespaces.
+	// This field is required.
+	//
+	// Example: Selects Pods with label "app" == "logging" in all Namespaces and
+	//          Pods with label "app" == "prometheus" in the "monitoring"
+	//          Namespace.
+	//
+	//  - selector:
+	//      matchExpressions:
+	//      - key: app
+	//        operator: In
+	//        values:
+	//        - logging
+	//  - selector:
+	//      matchLabels:
+	//        app: prometheus
+	//    namespaceSelector:
+	//      matchLabels:
+	//        kubernetes.io/metadata.name: monitoring
+	//
+	// +required
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:MaxItems=32
+	Pods []MachineDrainRulePodSelector `json:"pods"`
+}
+
+// MachineDrainRuleDrainConfig configures if and how Pods are drained.
+type MachineDrainRuleDrainConfig struct {
+	// behavior defines the drain behavior.
+	// Can be either "Drain" or "Skip".
+	// "Drain" means that the Pods to which this MachineDrainRule applies will be drained.
+	// If behavior is set to "Drain" the order in which Pods are drained can be configured
+	// with the order field. When draining Pods of a Node the Pods will be grouped by order
+	// and one group after another will be drained (by increasing order). Cluster API will
+	// wait until all Pods of a group are terminated / removed from the Node before starting
+	// with the next group.
+	// "Skip" means that the Pods to which this MachineDrainRule applies will be skipped during drain.
+	// +required
+	Behavior MachineDrainRuleDrainBehavior `json:"behavior"`
+
+	// order defines the order in which Pods are drained.
+	// Pods with higher order are drained after Pods with lower order.
+	// order can only be set if behavior is set to "Drain".
+	// If order is not set, 0 will be used.
+	// Valid values for order are from -2147483648 to 2147483647 (inclusive).
+	// +optional
+	Order *int32 `json:"order,omitempty"`
+}
+
+type MachineDrainRuleMachineSelector struct { //nolint:revive // Intentionally not adding a godoc comment as it would show up additionally to the field comment in the CRD
```

Comment:
I think on this, and the other selector, we want to add a `MinProperties` to ensure the struct isn't empty.

Why? Because otherwise I think we may end up being able to express the same intent in many ways

The fields for machines and pods have documentation that infers they are optional, but are currently required and have min properties 1, so, someone who wants to select all machines in all clusters has 1 way presently to represent this
```
machines:
- {}
```

Great!

What if we now change any of those other properties? Make the field optional, remove the minItems, now I could say

```
machines: null
```
or
```
machines: []
```
or
```
machines:
- {}
```
To all mean the same thing, which is confusing

So once we clarify the intent of the select all behaviour, we need to revisit this.

I suspect we want the `machines` and `pods` to be optional, and to have `minItems: 1`, so that `machines: []` is not valid and `machines: [{}]` is also not valid, but `machines: null` is used to represent no choice

Reply:
The original reason why I made `machines` required was to require an explicit opt-int if someone wants to apply a MachineDrainRule to all Pods or all Machines. But sounds like that doesn't work well

Reply:
Also the general standard for single selector fields seems to be: optional, omitempty, select all if not set

Reply:
> The original reason why I made machines required was to require an explicit opt-int if someone wants to apply a MachineDrainRule to all Pods or all Machines. But sounds like that doesn't work well

This _is _ possible, required field, minItems: 0, and don't let the fields within have 0 properties. *But* I think people are more used to omitting completely, and may not realise that a required `machines` means that they can do `machines: []`

Reply:
Yeah let's go with https://github.com/kubernetes-sigs/cluster-api/pull/11353#discussion_r1824392367

Reply:
Done